### DISCUSSION TITLE
Add helix to guix

### DISCUSSION THREAD

**Discussion by crmsnbleyd** [helix-editor/helix#10423] [General]
[Guix](https://guix.gnu.org/) is a package manager/operating system quite like Nix.

Currently, there isn't any package for the helix editor in the official channel, and I'm working on adding it.
I will respond here when it's available. 
Any help from guix users would also be appreciated, since this is my first time packaging a rust project from scratch.

**Comment by adamantinum** [comment]
I haven't used Guix in a while, but I found [this](https://guix.gnu.org/cookbook/en/html_node/Packaging-Tutorial.html) tutorial about packaging. I guess you should use `#:use-module (guix build-system cargo)` and then `(build-system cargo-build-system)`.

To get the source I think the best is to use `#:use-module (guix git-download)`.

**Comment by crmsnbleyd** [comment]
you are quite correct. Guix actually has an in-built method to package crates, using `guix import crate`.

However, since this project does not actually publish anything to crates.io yet is still divided into multiple crates,
I do not know how to organize the package. I have to package helix-{core,term,etc} and then the editor itself.
I have asked someone what I should do in this case and I expect they will respond soon. 

After that, just a bit of elbow grease, and you can expect my work to be done.

**Comment by alxsimon** [comment]
Hi, any news on this? I'd be curious to see a first attempt at this if you have one. I can also try playing with a draft package and see if that works for me.

**Comment by crmsnbleyd** [comment]
Hello, I have not been able to make any progress because I have not gotten
advice on the correct way to package it, if you're willing to write up a
draft (if that's what you meant) I can help with patching paths and so on.
If you just wish to use helix, a binary package is available on some
channels.

**Comment by alxsimon** [comment]
Unfortunately I'm too much of a beginner with guix to try and package it. But I would like to start packaging other software of the same type written in rust and wanted to see an example (as well as using helix). I'll have a look for binary packages thanks.

**Comment by crmsnbleyd** [comment]
You can take a look at just, which I packaged.

https://packages.guix.gnu.org/packages/just/1.14.0/

**Comment by lucyColeclough** [comment]
this has been talked about on the mailing list for a bit of time
one thread was mentioned today but is not on the archive yet but here is one:
https://lists.gnu.org/archive/html/guix-devel/2023-11/msg00006.html

**Comment by bhootd** [comment]
[This guix channel](https://codeberg.org/unmatched-paren/guix-channel) already has packaged [helix](https://codeberg.org/unmatched-paren/guix-channel/src/branch/root/paren/packages/helix.scm) (albeit two years ago). Its package description can perhaps help.

**Comment by ghost** [comment]
I [packaged latest master](https://codeberg.org/look/saayix/src/commit/f15663438a4c0d595b4a4f64e089d0fefd82a1c5/saayix/packages/rust/helix.scm) (00e9e5eadef16dd20cd24d303a664faaeb8faa56) for Guix with most of the grammars, haven't investigated why a handful don't pass the pack phase, I suspect bad file names in the store(?).
I don't use helix (in fact I didn't even know the project existed until yesterday) so I can't really test, but `hx --health` seems good to me.
The package definition should be self-contained, which means it doesn't use any rust package from the (gnu packages) module, so you can pretty much migrate it to anywhere easily.
You can also use my [channel](https://codeberg.org/look/saayix#usage) if you prefer, the package is in (saayix packages text-editors).

**Reply by bhootd** [reply]
That is some impressive work! The package seems to work well in preliminary testing.

If anyone wants to use helix in guix, this package is the best bet.

**Reply by woshilapin** [reply]
@Muril-o Thank you very much for this work. Note that I added your channel to [https://toys.whereis.%E3%81%BF%E3%82%93%E3%81%AA/](https://toys.whereis.xn--q9jyb4c/) to improve discoverability.

**Reply by ghost** [reply]
Thanks, I'll see if I can upstream it to guix in a reasonable time, there are still some minor issues with the packaging I need to sort out until then. I'll probably wait until plugin manager fully releases because it might need some thinkering for reproducibility, and I might write something for managing helix plugins with guix too. I currently dislike the thought of having too much state to manage manually outside of guix. Would be really cool if the parametrized packages implementation is merged by then.

**Reply by lucyColeclough** [reply]
Thanks for this, one thing i am thinking is to use guix tree sitter builds for this and remove the bundled ones in build
Then a helix service ( for system and home) can create a runtime folder using the runtime output of a helix package combined with guix built tree sitter grammars. The service can have a list of default guix tree sitter grammar packages.
Also fun that helix uses scheme, was thinking it might be possible to pass the helix config as a s-exp

**Reply by ghost** [reply]
> one thing i am thinking is to use guix tree sitter builds for this and remove the bundled ones in build

Yes, I was working on this couple days ago and made a package without the grammar in build for myself to use, [here's a link to it](https://codeberg.org/look/saayix/src/commit/d8699c7d6bab9c9612a4da3b3563452095671a62/saayix/packages/rust/helix.scm#L7512-L7566). This way I can manage grammars with `hx --grammar` and have a clean helix runtime without guix intervention for now, until I figure out exactly how I wanna integrate the guix tree-sitter module with the runtime.

> Then a helix service ( for system and home) can create a runtime folder using the runtime output of a helix package combined with guix built tree sitter grammars. The service can have a list of default guix tree sitter grammar packages.

I haven't given much thought into how would be the best approach, but this seems to be the most promising idea, we could also use the service later for managing plugins and config too.

> Also fun that helix uses scheme, was thinking it might be possible to pass the helix config as a s-exp

Yes, it will definitely be possible, just gotta wait until the transition from toml to scheme is complete :pray:
